libraries/standard/mqtt/src/mqtt_state.c

```diff
@@ -425,7 +421,7 @@ MQTTPublishState_t MQTT_UpdateStatePublish( MQTTContext_t * pMqttContext,
                                     &foundQoS,
                                     &currentState );
 
-        if( foundQoS != qos )
+        if( ( recordIndex == MQTT_STATE_ARRAY_MAX_COUNT ) || ( foundQoS != qos ) )
```

Comment:
Adding a comment here about why we can just rely on the foundQoS instead of the returned index  would help readability.